bindings/rust/s2n-tls/src/renegotiate.rs

```diff
@@ -0,0 +1,677 @@
+            assert_eq!(n, 0, "Unexpected application data");
+            result = r;
+        }
+        unwrap_poll(result).expect("Renegotiate");
+        assert_eq!(wake_count, count_per_handshake * 2);
+
+        Ok(())
+    }
+
+    #[test]
+    fn do_renegotiate_with_async_init() -> Result<(), Box<dyn Error>> {
+        // To test that the initializer method triggers again on the second
+        // handshake, we need to set an easily verified connection-level value.
+        // The server name is convenient.
+        #[derive(Clone)]
+        struct TestInitializer {
+            count: usize,
+            server_name: String,
+        }
+        impl ConnectionInitializer for TestInitializer {
+            fn initialize_connection(
+                &self,
+                _: &mut crate::connection::Connection,
+            ) -> ConnectionFutureResult {
+                Ok(Some(Box::pin(self.clone())))
+            }
+        }
+        impl ConnectionFuture for TestInitializer {
+            fn poll(
+                self: Pin<&mut Self>,
+                conn: &mut Connection,
+                ctx: &mut core::task::Context,
+            ) -> Poll<Result<(), crate::error::Error>> {
+                ctx.waker().wake_by_ref();
+                let this = self.get_mut();
+                if this.count > 1 {
+                    // Repeatedly block the handshake in order to verify
+                    // that renegotiate can handle Pending callbacks.
+                    this.count = this.count - 1;
+                    Pending
+                } else {
+                    conn.set_server_name(&this.server_name)?;
+                    Ready(Ok(()))
+                }
+            }
+        }
+
+        let count_per_handshake = 10;
+        let expected_server_name = "helloworld";
+        let initializer = TestInitializer {
+            count: count_per_handshake,
+            server_name: expected_server_name.to_owned(),
+        };
+
+        let mut builder = config::Builder::new();
+        builder.set_renegotiate_callback(RenegotiateResponse::Accept)?;
+        builder.set_connection_initializer(initializer)?;
+        let mut pair = RenegotiateTestPair::from(builder)?;
+
+        let (waker, wake_count) = new_count_waker();
+        pair.client.set_waker(Some(&waker))?;
+
+        pair.handshake().expect("Initial handshake");
+        assert_eq!(wake_count, count_per_handshake);
+        pair.trigger_renegotiate().expect("Trigger renegotiate");
+        pair.send_and_receive()
+            .expect("Application data before renegotiate");
+        pair.client
+            .wipe_for_renegotiate()
+            .expect("Wipe for renegotiate");
+        // Verify that the wipe cleared the server name
+        assert!(pair.client.server_name().is_none());
+
+        let mut empty = [0; 0];
+        let mut buf = [0; 1];
+        let mut result = Pending;
+        while result.is_pending() {
```

Comment:
It looks like these while loops are shared structure between 3 of the unit tests, could we maybe try and abstract that some of this out into a helper function?